### ISSUE TITLE
Extending base shop model vs moving functions to a trait

### ISSUE THREAD

**Issue by awebartisan** [ohmybrew/laravel-shopify#137]
Hi,
I am using this package and extending the BaseShop, I need to use another package which requires model to extend their base class too. How we are going to deal with such situation? Can we move the code in Shop.php model of this package to a trait?

Please have a look into it. thank you 👍

**Comment by ohmybrew** [comment]
I can move the code to a trait yes and leave the default model still there including the trait for backwards compatibility.

**Comment by awebartisan** [comment]
So if someone use only trait, what will happen to ShopObserver? I can change to the model to `\App\Shop` in the config. ShopObserver will not fire and it will cause trouble. What are your thoughts on this one?

**Comment by ohmybrew** [comment]
The Observer grabs the class defined in the config. It will be fine.

**Comment by awebartisan** [comment]
`use OhMyBrew\ShopifyApp\Models\Shop;`
Observer has this use statement on its top and it uses this in `creating` function. I have tried moving all the code from BaseShop model of package and moved to my own Shop model and removed extends keyword then i faced ShopObserver problem.

**Comment by ohmybrew** [comment]
I addressed this a couple versions back I believe already, someone made a
PR to use the config version.

**Comment by awebartisan** [comment]
```        // Shop observer
        $shopModel = config('shopify-app.shop_model');
        $shopModel::observe(ShopObserver::class);
```
Yes they did this in `ShopifyAppProvider.php` , but `ShopObserver.php` class still uses package base model in `use` statement and `creating` function.

**Comment by ohmybrew** [comment]
Hmm it just states it has to be an instance of Shop, not use it. It'll have
to be killed for the trait yes.

**Comment by ohmybrew** [comment]
It turns out this may not be possible.

I can not have $fillable, $casts, $dates, etc properties inside the trait. Because the base Eloquent defines them, the trait can not override it. Moving the properties and defining them inside the constructor of the model does not work, I get "fillable" errors on a few tests.

There does not seem to be a work around for this unfortunately. Only work around I see is to define all methods in the trait, and keep the properties inside the model class.

This is not ideal, but it would mean if you use the trait directly (without extending the shop model), you would need to also manually add the properties to your own class to ensure it works.

On the fence about this one...

**Comment by awebartisan** [comment]
I am also searching, will let you know if I get some idea from anywhere. Looking at other packages specifically which require extending their models for overriding methods and properties.

**Comment by awebartisan** [comment]
[https://gist.github.com/JordanDalton/f952b053ef188e8750177bf0260ce166](https://gist.github.com/JordanDalton/f952b053ef188e8750177bf0260ce166)

you probably stumbled upon this also?

**Comment by ohmybrew** [comment]
I have but this involves overwriting how eloquent does fillables and such. Which would depend on me updating the code should that change. Doesn't seem like a clean solution can happen besides the one I mentioned above which is still not totally clean but at least helps anyone wanting to extend

**Comment by awebartisan** [comment]
Okay... can you add the trait including only functions and mention in documentation to use it on your own risk? and add the properties to your own model.

**Comment by ohmybrew** [comment]
Yup that should work

**Comment by awebartisan** [comment]
Any timeline about this feature?

**Comment by awebartisan** [comment]
[https://www.archybold.com/blog/post/booting-eloquent-model-traits](https://www.archybold.com/blog/post/booting-eloquent-model-traits)

Does this help in any way to solve above problem?

**Comment by ohmybrew** [comment]
Sadly the above still causes complaints by Eloquent, it just wont work nicely.

I've gone the trait + model for properties. Its now available through v4.2.0 release. All tests passed, code coverage remained at 100.

See the wiki article on override and extending for more information; essentially use the trait and paste in the same properties from the base model.

**Comment by ohmybrew** [comment]
Observer will now accept any object yes.
